LLVM Build fails if exception enabled

CMake options to reproduce.
`-DLLVM_ENABLE_EH=ON -DLLVM_ENABLE_RTTI=ON -DLLVM_BUILD_EXAMPLES=ON`
```console
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp: In function ‘uintptr_t readSLEB128(const uint8_t**)’:
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:374:22: warning: left shift of negative value [-Wshift-negative-value]
  374 |     result |= (~0 << shift);
      |                      ^~~~~
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp: In function ‘llvm::BasicBlock* createFinallyBlock(llvm::LLVMContext&, llvm::Module&, llvm::IRBuilder<>&, llvm::Function&, std::string&, std::string&, llvm::BasicBlock&, llvm::BasicBlock&, llvm::Value**, llvm::Value**, llvm::Value**)’:
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:1001:76: error: no matching function for call to ‘llvm::IRBuilder<>::CreateLoad(llvm::Value*&)’
 1001 |                                                        *exceptionCaughtFlag),
      |                                                                            ^
In file included from /home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:56:
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1735:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, const char*)’
 1735 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, const char *Name) {
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1735:13: note:   candidate expects 3 arguments, 1 provided
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1739:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, const llvm::Twine&)’
 1739 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, const Twine &Name = "") {
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1739:13: note:   candidate expects 3 arguments, 1 provided
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1743:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, bool, const llvm::Twine&)’
 1743 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, bool isVolatile,
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1743:13: note:   candidate expects 4 arguments, 1 provided
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp: In function ‘llvm::Function* createCatchWrappedInvokeFunction(llvm::Module&, llvm::IRBuilder<>&, llvm::legacy::FunctionPassManager&, llvm::Function&, std::string, unsigned int, unsigned int*)’:
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:1189:57: error: no matching function for call to ‘llvm::IRBuilder<>::CreateLoad(llvm::Value*&)’
 1189 |                      builder.CreateLoad(exceptionStorage));
      |                                                         ^
In file included from /home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:56:
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1735:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, const char*)’
 1735 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, const char *Name) {
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1735:13: note:   candidate expects 3 arguments, 1 provided
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1739:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, const llvm::Twine&)’
 1739 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, const Twine &Name = "") {
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1739:13: note:   candidate expects 3 arguments, 1 provided
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1743:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, bool, const llvm::Twine&)’
 1743 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, bool isVolatile,
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1743:13: note:   candidate expects 4 arguments, 1 provided
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:1209:62: error: no matching function for call to ‘llvm::IRBuilder<>::CreateLoad(llvm::Value*&)’
 1209 |   builder.CreateResume(builder.CreateLoad(caughtResultStorage));
      |                                                              ^
In file included from /home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:56:
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1735:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, const char*)’
 1735 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, const char *Name) {
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1735:13: note:   candidate expects 3 arguments, 1 provided
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1739:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, const llvm::Twine&)’
 1739 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, const Twine &Name = "") {
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1739:13: note:   candidate expects 3 arguments, 1 provided
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1743:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, bool, const llvm::Twine&)’
 1743 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, bool isVolatile,
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1743:13: note:   candidate expects 4 arguments, 1 provided
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:1249:13: error: no matching function for call to ‘llvm::IRBuilder<>::CreateLoad(llvm::Value*)’
 1249 |           0));
      |             ^
In file included from /home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:56:
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1735:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, const char*)’
 1735 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, const char *Name) {
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1735:13: note:   candidate expects 3 arguments, 1 provided
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1739:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, const llvm::Twine&)’
 1739 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, const Twine &Name = "") {
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1739:13: note:   candidate expects 3 arguments, 1 provided
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1743:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, bool, const llvm::Twine&)’
 1743 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, bool isVolatile,
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1743:13: note:   candidate expects 4 arguments, 1 provided
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:1282:79: error: no matching function for call to ‘llvm::IRBuilder<>::CreateConstGEP1_64(llvm::Value*&, int64_t&)’
 1282 |                                                        ourBaseFromUnwindOffset),
      |                                                                               ^
In file included from /home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:56:
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1872:10: note: candidate: ‘llvm::Value* llvm::IRBuilderBase::CreateConstGEP1_64(llvm::Type*, llvm::Value*, uint64_t, const llvm::Twine&)’
 1872 |   Value *CreateConstGEP1_64(Type *Ty, Value *Ptr, uint64_t Idx0,
      |          ^~~~~~~~~~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1872:10: note:   candidate expects 4 arguments, 2 provided
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:1298:63: error: no matching function for call to ‘llvm::IRBuilder<>::CreateLoad(llvm::Value*&)’
 1298 |                        *(builder.CreateLoad(typeInfoThrownType)),
      |                                                               ^
In file included from /home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:56:
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1735:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, const char*)’
 1735 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, const char *Name) {
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1735:13: note:   candidate expects 3 arguments, 1 provided
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1739:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, const llvm::Twine&)’
 1739 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, const Twine &Name = "") {
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1739:13: note:   candidate expects 3 arguments, 1 provided
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1743:13: note: candidate: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Type*, llvm::Value*, bool, const llvm::Twine&)’
 1743 |   LoadInst *CreateLoad(Type *Ty, Value *Ptr, bool isVolatile,
      |             ^~~~~~~~~~
/home/kkumar/llvm-project/llvm/include/llvm/IR/IRBuilder.h:1743:13: note:   candidate expects 4 arguments, 1 provided
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp: In function ‘void runExceptionThrow(llvm::ExecutionEngine*, llvm::Function*, int32_t)’:
/home/kkumar/llvm-project/llvm/examples/ExceptionDemo/ExceptionDemo.cpp:1563:29: warning: catching polymorphic type ‘class {anonymous}::OurCppRunException’ by value [-Wcatch-value=]
 1563 |   catch (OurCppRunException exc) {
      |                             ^~~
```

--- a/llvm/examples/ExceptionDemo/CMakeLists.txt
+++ b/llvm/examples/ExceptionDemo/CMakeLists.txt
@@ -1,9 +1,7 @@
 set(LLVM_LINK_COMPONENTS
   Core
   ExecutionEngine
-  MC
-  MCJIT
-  RuntimeDyld
+  ORCJIT
   Support
   Target
   nativecodegen

--- a/llvm/examples/ExceptionDemo/ExceptionDemo.cpp
+++ b/llvm/examples/ExceptionDemo/ExceptionDemo.cpp
@@ -1277,8 +1277,8 @@ static llvm::Function *createCatchWrappedInvokeFunction(
   // (OurException instance).
   //
   // Note: ourBaseFromUnwindOffset is usually negative
-  llvm::Value *typeInfoThrown = builder.CreateConstGEP1_64(unwindException,
-                                                      ourBaseFromUnwindOffset));
+  llvm::Value *typeInfoThrown = builder.CreateConstGEP1_64(builder.getPtrTy(), unwindException,
+                                                           ourBaseFromUnwindOffset);
 
   // Retrieve thrown exception type info type
   //
@@ -1289,10 +1289,9 @@ static llvm::Function *createCatchWrappedInvokeFunction(
   llvm::Value *typeInfoThrownType =
       builder.CreateStructGEP(ourTypeInfoType, typeInfoThrown, 0);
 
-  llvm::Value *ti8 =
-      builder.CreateLoad(builder.getInt8Ty(), typeInfoThrownType);
-  generateIntegerPrint(context, module, builder, *toPrint32Int,
-                       builder.CreateZExt(ti8, builder.getInt32Ty()),
+  llvm::Value *ti32 =
+      builder.CreateLoad(builder.getInt32Ty(), typeInfoThrownType);
+  generateIntegerPrint(context, module, builder, *toPrint32Int, ti32,
                        "Gen: Exception type <%d> received (stack unwound) "
                        " in " +
                            ourId + ".\n",

--- a/llvm/examples/ExceptionDemo/ExceptionDemo.cpp
+++ b/llvm/examples/ExceptionDemo/ExceptionDemo.cpp
@@ -49,8 +49,9 @@
 
 #include "llvm/ADT/STLExtras.h"
 #include "llvm/BinaryFormat/Dwarf.h"
-#include "llvm/ExecutionEngine/MCJIT.h"
-#include "llvm/ExecutionEngine/SectionMemoryManager.h"
+#include "llvm/ExecutionEngine/Orc/Core.h"
+#include "llvm/ExecutionEngine/Orc/JITTargetMachineBuilder.h"
+#include "llvm/ExecutionEngine/Orc/LLJIT.h"
 #include "llvm/IR/DataLayout.h"
 #include "llvm/IR/DerivedTypes.h"
 #include "llvm/IR/IRBuilder.h"
@@ -83,6 +84,8 @@
 #ifndef USE_GLOBAL_STR_CONSTS
 #define USE_GLOBAL_STR_CONSTS true
 #endif
+
+llvm::ExitOnError ExitOnErr;
 
 //
 // Example types
@@ -142,6 +145,7 @@ static llvm::ConstantInt *ourExceptionCaughtState;
 
 typedef std::vector<std::string> ArgNames;
 typedef std::vector<llvm::Type*> ArgTypes;
+typedef llvm::ArrayRef<llvm::Type *> TypeArray;
 
 //
 // Code Generation Utilities
@@ -891,13 +895,10 @@ void generateStringPrint(llvm::LLVMContext &context,
 ///        generated, and is used to hold the constant string. A value of
 ///        false indicates that the constant string will be stored on the
 ///        stack.
-void generateIntegerPrint(llvm::LLVMContext &context,
-                          llvm::Module &module,
+void generateIntegerPrint(llvm::LLVMContext &context, llvm::Module &module,
                           llvm::IRBuilder<> &builder,
-                          llvm::Function &printFunct,
-                          llvm::Value &toPrint,
-                          std::string format,
-                          bool useGlobal = true) {
+                          llvm::Function &printFunct, llvm::Value *toPrint,
+                          std::string format, bool useGlobal = true) {
   llvm::Constant *stringConstant =
     llvm::ConstantDataArray::getString(context, format);
   llvm::Value *stringVar;
@@ -919,9 +920,8 @@ void generateIntegerPrint(llvm::LLVMContext &context,
 
   llvm::Value *cast = builder.CreateBitCast(stringVar,
                                             builder.getPtrTy());
-  builder.CreateCall(&printFunct, {&toPrint, cast});
+  builder.CreateCall(&printFunct, {toPrint, cast});
 }
-
 
 /// Generates code to handle finally block type semantics: always runs
 /// regardless of whether a thrown exception is passing through or the
@@ -996,10 +996,10 @@ static llvm::BasicBlock *createFinallyBlock(llvm::LLVMContext &context,
                       bufferToPrint.str(),
                       USE_GLOBAL_STR_CONSTS);
 
-  llvm::SwitchInst *theSwitch = builder.CreateSwitch(builder.CreateLoad(
-                                                       *exceptionCaughtFlag),
-                                                     &terminatorBlock,
-                                                     2);
+  llvm::SwitchInst *theSwitch = builder.CreateSwitch(
+      builder.CreateLoad(ourExceptionNotThrownState->getType(),
+                         *exceptionCaughtFlag),
+      &terminatorBlock, 2);
   theSwitch->addCase(ourExceptionCaughtState, &terminatorBlock);
   theSwitch->addCase(ourExceptionThrownState, &unwindResumeBlock);
 
@@ -1185,7 +1185,7 @@ static llvm::Function *createCatchWrappedInvokeFunction(
 
   // Note: function handles NULL exceptions
   builder.CreateCall(deleteOurException,
-                     builder.CreateLoad(exceptionStorage));
+                     builder.CreateLoad(builder.getPtrTy(), exceptionStorage));
   builder.CreateRetVoid();
 
   // Normal Block
@@ -1205,7 +1205,8 @@ static llvm::Function *createCatchWrappedInvokeFunction(
 
   builder.SetInsertPoint(unwindResumeBlock);
 
-  builder.CreateResume(builder.CreateLoad(caughtResultStorage));
+  builder.CreateResume(
+      builder.CreateLoad(ourCaughtResultType, caughtResultStorage));
 
   // Exception Block
 
@@ -1240,8 +1241,9 @@ static llvm::Function *createCatchWrappedInvokeFunction(
   // Retrieve exception_class member from thrown exception
   // (_Unwind_Exception instance). This member tells us whether or not
   // the exception is foreign.
-  llvm::Value *unwindExceptionClass =
-      builder.CreateLoad(builder.CreateStructGEP(
+  llvm::Value *unwindExceptionClass = builder.CreateLoad(
+      builder.getInt64Ty(),
+      builder.CreateStructGEP(
           ourUnwindExceptionType,
           unwindException,
           0));
@@ -1285,17 +1287,15 @@ static llvm::Function *createCatchWrappedInvokeFunction(
   typeInfoThrown = builder.CreateStructGEP(ourExceptionType, typeInfoThrown, 0);
 
   llvm::Value *typeInfoThrownType =
-      builder.CreateStructGEP(builder.getPtrTy(), typeInfoThrown, 0);
+      builder.CreateStructGEP(ourTypeInfoType, typeInfoThrown, 0);
 
-  generateIntegerPrint(context,
-                       module,
-                       builder,
-                       *toPrint32Int,
-                       *(builder.CreateLoad(typeInfoThrownType)),
+  llvm::Value *ti8 =
+      builder.CreateLoad(builder.getInt8Ty(), typeInfoThrownType);
+  generateIntegerPrint(context, module, builder, *toPrint32Int,
+                       builder.CreateZExt(ti8, builder.getInt32Ty()),
                        "Gen: Exception type <%d> received (stack unwound) "
                        " in " +
-                       ourId +
-                       ".\n",
+                           ourId + ".\n",
                        USE_GLOBAL_STR_CONSTS);
 
   // Route to matched type info catch block or run cleanup finally block
@@ -1307,8 +1307,7 @@ static llvm::Function *createCatchWrappedInvokeFunction(
 
   for (unsigned i = 1; i <= numExceptionsToCatch; ++i) {
     nextTypeToCatch = i - 1;
-    switchToCatchBlock->addCase(llvm::ConstantInt::get(
-                                   llvm::Type::getInt32Ty(context), i),
+    switchToCatchBlock->addCase(llvm::ConstantInt::get(builder.getInt32Ty(), i),
                                 catchBlocks[nextTypeToCatch]);
   }
 
@@ -1383,14 +1382,10 @@ createThrowExceptionFunction(llvm::Module &module, llvm::IRBuilder<> &builder,
   builder.SetInsertPoint(entryBlock);
 
   llvm::Function *toPrint32Int = module.getFunction("print32Int");
-  generateIntegerPrint(context,
-                       module,
-                       builder,
-                       *toPrint32Int,
-                       *exceptionType,
-                       "\nGen: About to throw exception type <%d> in " +
-                       ourId +
-                       ".\n",
+  generateIntegerPrint(context, module, builder, *toPrint32Int,
+                       builder.CreateZExt(exceptionType, builder.getInt32Ty()),
+                       "\nGen: About to throw exception type <%d> in " + ourId +
+                           ".\n",
                        USE_GLOBAL_STR_CONSTS);
 
   // Switches on runtime type info type value to determine whether or not
@@ -1542,15 +1537,13 @@ typedef void (*OurExceptionThrowFunctType) (int32_t typeToThrow);
 /// @param function generated test function to run
 /// @param typeToThrow type info type of generated exception to throw, or
 ///        indicator to cause foreign exception to be thrown.
-static
-void runExceptionThrow(llvm::ExecutionEngine *engine,
-                       llvm::Function *function,
-                       int32_t typeToThrow) {
+static void runExceptionThrow(llvm::orc::LLJIT *JIT, std::string function,
+                              int32_t typeToThrow) {
 
   // Find test's function pointer
   OurExceptionThrowFunctType functPtr =
-    reinterpret_cast<OurExceptionThrowFunctType>(
-       reinterpret_cast<intptr_t>(engine->getPointerToFunction(function)));
+      reinterpret_cast<OurExceptionThrowFunctType>(reinterpret_cast<uintptr_t>(
+          ExitOnErr(JIT->lookup(function)).getValue()));
 
   try {
     // Run test
@@ -1578,8 +1571,6 @@ void runExceptionThrow(llvm::ExecutionEngine *engine,
 //
 // End test functions
 //
-
-typedef llvm::ArrayRef<llvm::Type*> TypeArray;
 
 /// This initialization routine creates type info globals and
 /// adds external function declarations to module.
@@ -1890,93 +1881,73 @@ int main(int argc, char *argv[]) {
     return(0);
   }
 
-  // If not set, exception handling will not be turned on
-  llvm::TargetOptions Opts;
-
   llvm::InitializeNativeTarget();
   llvm::InitializeNativeTargetAsmPrinter();
-  llvm::LLVMContext Context;
-  llvm::IRBuilder<> theBuilder(Context);
+  auto Context = std::make_unique<llvm::LLVMContext>();
+  llvm::IRBuilder<> theBuilder(*Context);
 
   // Make the module, which holds all the code.
   std::unique_ptr<llvm::Module> Owner =
-      std::make_unique<llvm::Module>("my cool jit", Context);
+      std::make_unique<llvm::Module>("my cool jit", *Context);
   llvm::Module *module = Owner.get();
 
-  std::unique_ptr<llvm::RTDyldMemoryManager> MemMgr(new llvm::SectionMemoryManager());
+  // Build LLJIT
+  std::unique_ptr<llvm::orc::LLJIT> JIT =
+      ExitOnErr(llvm::orc::LLJITBuilder().create());
 
-  // Build engine with JIT
-  llvm::EngineBuilder factory(std::move(Owner));
-  factory.setEngineKind(llvm::EngineKind::JIT);
-  factory.setTargetOptions(Opts);
-  factory.setMCJITMemoryManager(std::move(MemMgr));
-  llvm::ExecutionEngine *executionEngine = factory.create();
+  // Set up the optimizer pipeline.
+  llvm::legacy::FunctionPassManager fpm(module);
 
-  {
-    llvm::legacy::FunctionPassManager fpm(module);
-
-    // Set up the optimizer pipeline.
-    // Start with registering info about how the
-    // target lays out data structures.
-    module->setDataLayout(executionEngine->getDataLayout());
-
-    // Optimizations turned on
+  // Optimizations turned on
 #ifdef ADD_OPT_PASSES
 
-    // Basic AliasAnslysis support for GVN.
-    fpm.add(llvm::createBasicAliasAnalysisPass());
+  // Basic AliasAnslysis support for GVN.
+  fpm.add(llvm::createBasicAliasAnalysisPass());
 
-    // Promote allocas to registers.
-    fpm.add(llvm::createPromoteMemoryToRegisterPass());
+  // Promote allocas to registers.
+  fpm.add(llvm::createPromoteMemoryToRegisterPass());
 
-    // Do simple "peephole" optimizations and bit-twiddling optzns.
-    fpm.add(llvm::createInstructionCombiningPass());
+  // Do simple "peephole" optimizations and bit-twiddling optzns.
+  fpm.add(llvm::createInstructionCombiningPass());
 
-    // Reassociate expressions.
-    fpm.add(llvm::createReassociatePass());
+  // Reassociate expressions.
+  fpm.add(llvm::createReassociatePass());
 
-    // Eliminate Common SubExpressions.
-    fpm.add(llvm::createGVNPass());
+  // Eliminate Common SubExpressions.
+  fpm.add(llvm::createGVNPass());
 
-    // Simplify the control flow graph (deleting unreachable
-    // blocks, etc).
-    fpm.add(llvm::createCFGSimplificationPass());
+  // Simplify the control flow graph (deleting unreachable
+  // blocks, etc).
+  fpm.add(llvm::createCFGSimplificationPass());
 #endif  // ADD_OPT_PASSES
 
-    fpm.doInitialization();
+  fpm.doInitialization();
 
-    // Generate test code using function throwCppException(...) as
-    // the function which throws foreign exceptions.
-    llvm::Function *toRun =
-    createUnwindExceptionTest(*module,
-                              theBuilder,
-                              fpm,
-                              "throwCppException");
+  // Generate test code using function throwCppException(...) as
+  // the function which throws foreign exceptions.
+  createUnwindExceptionTest(*module, theBuilder, fpm, "throwCppException");
 
-    executionEngine->finalizeObject();
+  ExitOnErr(JIT->addIRModule(
+      llvm::orc::ThreadSafeModule(std::move(Owner), std::move(Context))));
 
 #ifndef NDEBUG
-    fprintf(stderr, "\nBegin module dump:\n\n");
+  fprintf(stderr, "\nBegin module dump:\n\n");
 
-    module->dump();
+  module->print(llvm::errs(), nullptr);
 
-    fprintf(stderr, "\nEnd module dump:\n");
+  fprintf(stderr, "\nEnd module dump:\n");
 #endif
 
-    fprintf(stderr, "\n\nBegin Test:\n");
+  fprintf(stderr, "\n\nBegin Test:\n");
+  std::string toRun = "outerCatchFunct";
 
-    for (int i = 1; i < argc; ++i) {
-      // Run test for each argument whose value is the exception
-      // type to throw.
-      runExceptionThrow(executionEngine,
-                        toRun,
-                        (unsigned) strtoul(argv[i], NULL, 10));
-    }
-
-    fprintf(stderr, "\nEnd Test:\n\n");
+  for (int i = 1; i < argc; ++i) {
+    // Run test for each argument whose value is the exception
+    // type to throw.
+    runExceptionThrow(JIT.get(), toRun, (unsigned)strtoul(argv[i], NULL, 10));
   }
 
-  delete executionEngine;
+  fprintf(stderr, "\nEnd Test:\n\n");
 
   return 0;
 }

--- a/llvm/examples/ExceptionDemo/ExceptionDemo.cpp
+++ b/llvm/examples/ExceptionDemo/ExceptionDemo.cpp
@@ -1243,10 +1243,7 @@ static llvm::Function *createCatchWrappedInvokeFunction(
   // the exception is foreign.
   llvm::Value *unwindExceptionClass = builder.CreateLoad(
       builder.getInt64Ty(),
-      builder.CreateStructGEP(
-          ourUnwindExceptionType,
-          unwindException,
-          0));
+      builder.CreateStructGEP(ourUnwindExceptionType, unwindException, 0));
 
   // Branch to the externalExceptionBlock if the exception is foreign or
   // to a catch router if not. Either way the finally block will be run.
@@ -1277,8 +1274,8 @@ static llvm::Function *createCatchWrappedInvokeFunction(
   // (OurException instance).
   //
   // Note: ourBaseFromUnwindOffset is usually negative
-  llvm::Value *typeInfoThrown = builder.CreateConstGEP1_64(builder.getPtrTy(), unwindException,
-                                                           ourBaseFromUnwindOffset);
+  llvm::Value *typeInfoThrown = builder.CreateConstGEP1_64(
+      builder.getInt8Ty(), unwindException, ourBaseFromUnwindOffset);
 
   // Retrieve thrown exception type info type
   //
@@ -1541,8 +1538,7 @@ static void runExceptionThrow(llvm::orc::LLJIT *JIT, std::string function,
 
   // Find test's function pointer
   OurExceptionThrowFunctType functPtr =
-      reinterpret_cast<OurExceptionThrowFunctType>(reinterpret_cast<uintptr_t>(
-          ExitOnErr(JIT->lookup(function)).getValue()));
+      ExitOnErr(JIT->lookup(function)).toPtr<OurExceptionThrowFunctType>();
 
   try {
     // Run test
@@ -1851,7 +1847,8 @@ static void createStandardUtilityFunctions(unsigned numTypeInfos,
 
   // llvm.eh.typeid.for intrinsic
 
-  getDeclaration(&module, llvm::Intrinsic::eh_typeid_for, builder.getPtrTy());
+  getOrInsertDeclaration(&module, llvm::Intrinsic::eh_typeid_for,
+                         builder.getPtrTy());
 }
 
 

